pallene/checker.lua

```diff
@@ -765,19 +913,53 @@ function FunChecker:check_exp_synthesize(exp)
         local f_type = exp.exp._type
 
         if f_type._tag == "types.T.Function" then
+            local last_arg = exp.args[#exp.args]
+            for i = 1, #exp.args do
+                if f_type.arg_types[i] then
+                    exp.args[i] = self:check_exp_verify(exp.args[i],
+                                    f_type.arg_types[i],
+                                    "argument %d of call to function", i)
+                else
+                    type_error(exp.loc,
+                        "function expects %d argument(s) but received %d",
+                        #f_type.arg_types, #exp.args)
+                end
+            end
+            if last_arg and last_arg._types then
+                for i = 2, #last_arg._types do
+                    local narg = #f_type.arg_types - i + 1
+                    local param_type = f_type.arg_types[narg]
+                    local extra_exp = ast.Exp.ExtraRet(last_arg.loc, last_arg,
+                                            i, #last_arg._types)
+                    extra_exp._type = last_arg._types[i]
+                    if param_type then
+                        extra_exp = self:check_exp_verify(extra_exp, param_type,
+                                        "argument %d of call to function", i)
+                        table.insert(exp.args, extra_exp)
+                    else
+                        type_error(exp.loc,
+                            "function expects %d argument(s) but received %d",
+                            #f_type.arg_types, #exp.args)
+                    end
+                end
+
+            end
+
             if #f_type.arg_types ~= #exp.args then
                 type_error(exp.loc,
                     "function expects %d argument(s) but received %d",
                     #f_type.arg_types, #exp.args)
             end
-            for i = 1, math.min(#f_type.arg_types, #exp.args) do
-                exp.args[i] = self:check_exp_verify(
-                    exp.args[i], f_type.arg_types[i],
-                    "argument %d of call to function", i)
-            end
-            assert(#f_type.ret_types <= 1)
-            if #f_type.ret_types >= 1 then
+
+            if #f_type.ret_types == 1 or
+               (#f_type.ret_types > 1 and exp._single_ret)
+            then
                 exp._type = f_type.ret_types[1]
+
+            elseif #f_type.ret_types > 1 and not exp._single_ret then
+                exp._type  = f_type.ret_types[1]
+                exp._types = f_type.ret_types
```

Comment:
Would there be a problem if we got rid of the `_single_ret` and always set the `_types` for all CallFuncs?

Reply:
Currently this is used to know, in `to_ir`, if `ast.Exp.CallFunc` is the last call and we need to create the `ir.Cmd.CallStatic`. For example, if you write `return f()` and `f()` returns an integer, `_types` will not be set and this is the information we need in the node to see that we need to create the call in that evaluation. Same thing happens with a function call inside an `ast.Exp.Paren`. If we put `_types` in every function call we'll miss that information in `to_ir` and I couldn't think of another way to do that. This was the way I got it to work, but I understand that it is not very clean, especially in `ast.Exp.Paren`. I'm thinking about a new way to do this while I'm fixing the other issues that you pointed out.